serverlessworkflow/modules/ROOT/pages/core/understanding-jq-expressions.adoc

```diff
@@ -1 +1,153 @@
+
+image::core/expression_diagram.png[]
+
+This Serverless Workflow Application accepts as input an array of complex numbers (where `x` is the real coordinate and `y` the imaginary one) and defines an expression function (invoked by `squareState`) to calculate the maximum `x` and the minimum `y` for that `numbers` array. It also contains an action data filter (defined inside `squareState`) that selects the maximum `x` to be merged into the flow model and a state data filter (defined inside `finish` state) that sets that max value as the whole model that will be returned to the caller.
+
+
+.The expression function
+[source,json]
+----
+"functions": [
+    {
+      "name": "max",
+      "type": "expression",
+      "operation": "{max: .numbers | max_by(.x), min: .numbers | min_by(.y)}"
+    }
+]
+----
+
+We define a `max` function of type expression. The operation property is a string containing a `jq` expression. This expression returns a JSON object, where `max` property is the maximum value of `x` coordinate in the input array and `min` property is the minimum value of `y` coordinate in the same array. 
+
+.The action data filter
+[source,json]
+----
+"actions": [
+        {
+          "name": "maxAction",
+          "functionRef": {
+            "refName": "max"
+          },
+          "actionDataFilter": {
+             "results" : ".max.x",
+             "toStateData" : ".number"
+          }
+        }
+ ]
+----
+
+Since we are only interested in the maximum `x`, besides invoking the function using `functionRef`, this actions also contains an action data filter. If we were not adding this filter, the whole JSON Object returned by the function call will be merged into the flow model. The filter has two properties: `results`, which selects the attribute to be merged from the data returned by the action and `toStateData`, which indicates the name of the target property inside the flow model (in case this property does not exist, it will be added). So, after executing the action, the flow model will consist of a `number` property storing the maximum value and the original `numbers` array. Then the flow transitions to `finish` state.
+
+.The state data filter
+[source,json]
+----
+"name": "finish",
+"type": "operation",
+"stateDataFilter": {
+   "input": "{result: .number}"
+}
+----
+
+Since we do not want to return the original `numbers` array as a result of the flow execution, the final stage consists of a state data filter that sets the contents of the output model. Hence, we set the model to be a JSON object containing a property named `result`, whose value is the maximum number calculated by the previous state, which was stored in the `number` property. We achieve this by using the `input` property of the stateDataFilter construct, meaning that the model is changed before the state gets executed. As final result of this whole procedure, the  model content returned to the user contains a `result` property whose value is the maximum `x`.
```

Comment:
Can you add closing sections here? For example:

```asciidoc
== Additional resources

- {configuring-openapi-services-endpoints}[Configuring OpenAPI Services Endpoints]
- {camel-k-integration}[Camel K Integration]
- {authention-support-for-openapi-services}[Authentication Support for OpenAPI Services]

include::../../pages/_common-content/report-issue.adoc[]
```

Reply:
sure